File: cpp/90. subsets-ii.cpp
//
// Created by Rutvik Patel on 2/24/24.
//

#include "include.h"

// solution
class Solution {
private:
    void generateSubsets(const std::vector<int>& nums, int start, std::vector<int>& current, std::vector<std::vector<int>>& result) {
        result.push_back(current);
        for (int i = start; i < nums.size(); ++i) {
            if (i == start or nums[i] != nums[i - 1]) {
                current.push_back(nums[i]);
                generateSubsets(nums, i + 1, current, result);
                current.pop_back();
            }
        }
    }
public:
    std::vector<std::vector<int>> subsetsWithDup(std::vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        std::vector<std::vector<int>> result;
        std::vector<int> current;
        generateSubsets(nums, 0, current, result);
        return result;
    }
};

// main
int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::vector<int>, std::vector<std::vector<int>>>> test_cases = {
        { { 1, 2, 2 }, { { }, { 1 }, { 1, 2 }, { 1, 2, 2 }, { 2 }, { 2, 2 } } },
        { { 0 }, { { }, { 0 } } },
    };

    Solution solution;

    for(auto& [args, ans] : test_cases)
        assert_m(solution.subsetsWithDup(args) == ans, "Failed!");

    log("All testcases passed!");
}

File: python/90. subsets-ii.py
from itertools import chain, combinations
from typing import List


# lc sol start
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        return list(map(list, set(chain.from_iterable(combinations(nums, r) for r in range(len(nums) + 1)))))

    # the following solution is also correct and is written without using build in functions
    # def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
    #     res = [[]]
    #     nums.sort()
    #     for i in range(len(nums)):
    #         if i == 0 or nums[i] != nums[i - 1]:
    #             l = len(res)
    #         for j in range(len(res) - l, len(res)):
    #             res.append(res[j] + [nums[i]])
    #     return res


# lc sol end


def main() -> None:
    test_cases = {
        'inputs': [([1, 2, 2],), ([0],)],
        'outputs': [[[], [1], [2], [1, 2], [2, 2], [1, 2, 2]], [[], [0]]],
    }

    solution = Solution()

    for idx, args in enumerate(test_cases['inputs']):
        op = solution.subsetsWithDup(*args)
        assert sorted(op) == sorted(test_cases['outputs'][idx]), \
            f"expected output {test_cases['outputs'][idx]} for input {args}, got {op}"

    print('all sample test cases passed successfully!')


if __name__ == '__main__':
    main()
